### ISSUE TITLE
How to add your own voices?

### ISSUE THREAD

**Issue by zervankoru** [Olga-Yakovleva/RHVoice#95]
I just don't understand, how to add my own voices and create accoustic models, where to search for information?

**Comment by joshknnd1982** [comment]
I think that rh-voice would be perfect for preserving one of my favorite english-voice speech synthesizers which has not been available on windows for years. If I could have the required sentences that need to be recorded, I could have my favorite english-language tts, Keynote Gold, read each sentence into its own wave file then send you the recordings? then could you possibly make a Keynote gold voice, for rh-voice, for me? It would be a standard english-language voice. I would set keynote gold to read slowly so all phonemes could be captured. I would love to in the future have a windows-based application to build english-language voices in windows 10. and documentation for making new language voices would be great. thanks.

**Comment by joshknnd1982** [comment]
also, if its possible for me to send you a bunch of recordings of keynote gold, reading sentences, then hopefully the keynote gold rh-voice could be ported to android also?

**Comment by Olga-Yakovleva** [comment]
No, voices are built from recordings of natural speech, that is we 
record people, not other speech synthesizers.

And a graphical apllication for Windows which would allow anyone to 
create a new voice is not something I am going to work on. It would be 
too much work, and I don't think that I could create a program which 
would hide all the details from the user and would allow them to create 
a voice without understanding how the process works..

**Comment by joshknnd1982** [comment]
Ok then, perhaps you could write up a word document and describe how to build a voice when you get time? I would like to along with others, take up the challenge of making a keynote gold rh-voice, so that I can preserve my favorite tts from childhood. I think that rh-voice can do this because keynote gold was a very clear natural sounding tts engine for its time, best of its time in fact. 



Sent from Mail for Windows 10

**Comment by abitrolly** [comment]
@JulienCochuyt the link to wiki is 404.

**Comment by Olga-Yakovleva** [comment]
I need to think carefully how to describe the process. For example, I'm 
not sure what prior knowledge about speech synthesis and related 
subjects I can assume and what should be explained in my document. I 
can't write a tutorial on speech synthesis in general. Should it be just 
a list of commands to execute?

@abitrolly, do you mean that you will be following the instructions on 
your machine, performing all the steps to build a new voice?

@JulienCochuyt, compared to building a new voice for an already 
supported language, adding support for a new language requires much more 
work and much more documentation. I will only start thinking about it 
after I create a document about building a new voice which someone can 
actually use to build such a voice.

@abitrolly, I can't reproduce any problems with the wiki.

**Comment by beqabeqa473** [comment]
i have created a script which will set up an environment on a fresh os install.

i've tested it only on ubuntu 16.04, but it should work on other versions too.

On 11/19/18, Olga-Yakovleva <notifications@github.com> wrote:
> I need to think carefully how to describe the process. For example, I'm
> not sure what prior knowledge about speech synthesis and related
> subjects I can assume and what should be explained in my document. I
> can't write a tutorial on speech synthesis in general. Should it be just
> a list of commands to execute?
>
> @abitrolly, do you mean that you will be following the instructions on
> your machine, performing all the steps to build a new voice?
>
> @JulienCochuyt, compared to building a new voice for an already
> supported language, adding support for a new language requires much more
> work and much more documentation. I will only start thinking about it
> after I create a document about building a new voice which someone can
> actually use to build such a voice.
>
> @abitrolly, I can't reproduce any problems with the wiki.
>
>
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/Olga-Yakovleva/RHVoice/issues/95#issuecomment-439899288


-- 
with best regards beqa

**Comment by abitrolly** [comment]
@Olga-Yakovleva yes, I will follow the instructions. It would be nice if I can ask questions to expand the tutorial with details from theory.

**Comment by snakers4** [comment]
@Olga-Yakovleva 

Many thanks for amazing library (I could not find any donation details, though).
We are building an STT system.
One of problems of such systems is that they require a lot of data.
Sometimes creating artificial data helps (I am planning to launch a few experiments with 4 voices from your system as annotation for scarce data).

At the same time I have some access to the resources of several companies, which may be willing to invest a bit of time into this. They do not have enough resources to annotate / create 20k hours of STT annotation, but there maybe 20-30 people (or even maybe 100-200 if I ask a bit), who can donate their voices.

> No, voices are built from recordings of natural speech, that is we record people, not other speech synthesizers.

If I understand correctly, in such systems voices are created from voiced text / pronunciation n-grams.
If you could just provide the list of utterance to be voiced by people, I would do my best to provide recordings for at least several dozens of people.

**Comment by velmyshanovnyi** [comment]
Я готовий долучитись до укр мови, і підтягнути ще волонтерів.

Які саме тексти треба начитати/озвучити?

**Comment by Babdus** [comment]
Thank you for your effort for making this library and making it open to everyone.

I would like to add my voice as well. Although it is relatively old thread, I couldn't find another place for information about adding voice, nor did I found the doc or wiki mentioned above. So please execuse me and point me to the relevant page.

Otherwise, if it is possible, could you write down the steps I need to follow in order to have my voice?

I would like to have the new voice for Georgian. First question is what kind of utterances should I record and in what format should they be. Second, is there some kind of preprocessing needed, for example pitch normalization? Third, what sequence of commands should I run to get my voice model work?

Thanks in advance.

**Comment by velmyshanovnyi** [comment]
> Я готовий подати до укр мови, під час отримання ще волонтерів.
> 
> Які саме тексти потрібно читати / оглядати?

Я надиктовую/озвучую слова на www.forvo.com
Можу озвучити потрібну кількість слів Укр мови, і потім по алгоритму їх опрацювати.

Я айтішник, знання ffmpeg та архітектури програмування.

Що потрібно робити?

**Comment by velmyshanovnyi** [comment]
дорос до уровня создания своего голоса "Ant"

я так понял что для Украинского что нужно надиктовать:
https://github.com/Olga-Yakovleva/RHVoice/blob/master/src/scripts/Ukrainian/prompts/text1.txt

появляются вопросы:

- где почитать про механику и порядок создания?
- .wav ?
- 48кГц ?
- паузы перед/после, до скольки милисекунд?
- надиктовывась одним файлом, или резать?
- чистить по частотам, или просто записать?
- какие теги прописывать в exif ?

**Comment by velmyshanovnyi** [comment]
вопрос актуален, и не теряет своей актуальности!

**Comment by elnursamedzade** [comment]
I have my own recorded audio and text files.  Could you guide me to a new language?

**Comment by winman3000** [comment]
I am also still looking for how to build a new language and voice for RHVoice. Especially for German. Does anyone have a step by step guide?

Especially I am interested in the following questions:
1. What is needed to build a new voice and a new language, for example for German? I know that the text and the recording must exist.
1. How long must the recording be?
1. How exactly do I have to proceed if I have both the recording and the text?

I would be very happy to receive an answer. Unfortunately I have not found any documentation on this.

**Comment by KOLANICH** [comment]
>Unfortunately I have not found any documentation on this.

I guess that it may be intentional. There exists a business around making voices for RHVoice (and claiming copyright on them).

**Comment by winman3000** [comment]
I think I know what you mean. However, I would like to have a German voice for RHVoice. Unfortunately, I don't seem to get much help here...

**Comment by KOLANICH** [comment]
I guess there is not so many options except someone neutral and independent writing the docs oneself.

**Comment by zvezdochiot** [comment]
@winman3000 say:
>However, I would like to have a German voice for RHVoice. Unfortunately, I don't seem to get much help here...

Maybe use https://rhvoice.su/create/

**Comment by KOLANICH** [comment]
It won't solve the issue with lack of docs, instead it would reward the company for the lack of docs.

**Comment by nshmyrev** [comment]
@winman3000 you'd better check German voices here:

https://github.com/TensorSpeech/TensorFlowTTS

https://colab.research.google.com/drive/1W0nSFpsz32M0OcIkY9uMOiGrLTPKVhTy?usp=sharing

**Comment by zvezdochiot** [comment]
@KOLANICH say:
>It won't solve the issue with lack of docs, instead it would reward the company for the lack of docs.

Not? And what in return?

**Comment by alex19EP** [comment]
> 
> 
> > Unfortunately I have not found any documentation on this.
> 
> I guess that it may be intentional. There exists a business around making voices for RHVoice (and claiming copyright on them). A conflict of interests.

you're wrong the lack of documentation is unrelated to rhvoice lab and its voices.
this is only due to the lack of time on the part of the main developer.

> I think I know what you mean. However, I would like to have a German voice for RHVoice. Unfortunately, I don't seem to get much help here...

I'm really sorry, but Olga doesn't have a lot of free time now, and I'm far from being a specialist in these matters.

I can only say that creating new voices is much easier than creating a new language. for this you will definitely need knowledge of c ++ and hts technology

> I guess there is not so many options except someone neutral and independent writing the docs oneself.

I want to reiterate that the main engine remains an open source project under a free license.

> 
> 
> @winman3000 say:
> 
> > However, I would like to have a German voice for RHVoice. Unfortunately, I don't seem to get much help here...
> 
> Maybe use https://rhvoice.su/create/

I don't think RHVoice lab has people competent enough to create a new language.

> It won't solve the issue with lack of docs, instead it would reward the company for the lack of docs.

what company?

**Comment by abitrolly** [comment]
@winman3000 I suggest to create another issue specifically for German, because in HMM synthesis each language not only needs a voice data, but also a bit of programming for the "language model". Then if I had time, I would follow http://www.cs.columbia.edu/~ecooper/tts/data.html and report my progress there.

As for alternatives I'd expect the German model implemented in MaryTTS, but docs are archived and I can't see newer ones https://github.com/marytts/marytts-wiki/blob/master/HMMVoiceCreation.md @KOLANICH I suggest there is a global synthesizer voice conspiracy going on!

**Comment by winman3000** [comment]
@abitrolly: Thank you for your help. Unfortunately, I have no programming skills to program the German module. Too bad.

I already created an issue for it a few years ago, however this makes no progress. See Issue #24.

**Comment by KOLANICH** [comment]
>this is only due to the lack of time on the part of the main developer.

Yes. And I feel like that we shouldn't expect the lab staff (which I consider to be a part of this project too, so if they were not having a company making voices, I'd assigned to them the probabilities of creating the docs within a unit of time higher than most of contributors to this repo, but as they are members of an org, I currently assign them probabilities that are lower than other contributors have) to write the docs.

**Comment by winman3000** [comment]
@alex19EP:
> I don't think RHVoice lab has people competent enough to create a new language.

Then I'm afraid that there will never be a German voice for RHVoice, because unfortunately I lack the programming skills for it. Moreover, I don't know anyone who could help me here. Too bad, because RHVoice is very small and responsive.

Nevertheless, I thank you for your help!

**Comment by abitrolly** [comment]
@winman3000 @KOLANICH has programming skills. :D

**Comment by abitrolly** [comment]
I am trying to create my own voice for more than three years by now. I leaned I lot about HMM and signal processing, spent a lot of time talking about it and RHVoice, and yet I haven't found the time to just sit and follow the links I gave you to see how far I can get.

**Comment by abitrolly** [comment]
Yet the work done by Olga on RHVoice and Maria on MAGE to make the synthesis responsive is fascinating.

**Comment by winman3000** [comment]
@abitrolly: What I like about RHVoice is that it requires very few dependencies such as Java etc.. I would like to use it with NVDA, an open source screen reader for the blind. Unfortunately, since I lack programming experience as mentioned above and there doesn't seem to be a project to build on, I guess I'll have to pass on it.

Sorry for spamming this issue, I didn't mean to turn it into a discussion.

**Comment by winman3000** [comment]
Sorry, I have one last question about this:

is it possible to port the HMM-based voices from the Mary TTS project to RHVoice? Mary TTS has HMM-based German voices. Probably you need programming skills for this as well, right?

**Comment by maniyax** [comment]
Hi all!

1. RHVoice Lab may be create only voices, not language modules.
2. We haven't documentation of creating voices. Our activity is only experiment and experience. We don't have to write documentation.
3. We haven't resources for experiments of creating language modules. We waiting official documentation from Olga as everyone.
4. We don't the company or business organization. We just a team of blind specialists.

If you want create voice, you should send e-mail for us. If you want create a new language, you must wait official documentation.

**Comment by KOLANICH** [comment]
>We don't the company or business organization. We just a team of blind specialists.

Isn't a team already an organization?

https://en.wiktionary.org/wiki/organization

>A group of people consciously cooperating. 

A team surely matches this definition.

**Comment by KOLANICH** [comment]
>We don't the company or business organization. We just a team of blind specialists.

I wonder which entity holds the copyright on the voices the team produces and what are the usual conditions for using your voices. Do you require commercial users to pay?

**Comment by winman3000** [comment]
@maniyax: Sorry, it was not meant to be an accusation. I realize that you are a team of volunteers and you are not committed to anything. You have family, work, studies or whatever, and such things naturally take priority over volunteer projects.

Until recently, I didn't know that a new language would require a new language module and that programming skills would be required. Even if there would be a documentation later on how to develop a language module, I wouldn't be able to do anything with it anyway.

So I have no choice but to wait until someone takes on this problem and develops a german language module.

I will not comment on the other issues that have nothing to do with this topic.

**Comment by maniyax** [comment]
@KOLANICH 
1. A non-profit organization isn't a business organization.
2. NPO is official registrated legal entity. We don't organization.

**Comment by maniyax** [comment]
Hello, @winman3000!

When Olga writes documentation for creating new languages, we would be able  to start experiments.

**Comment by KOLANICH** [comment]
I have corrected my previous posts in this issue, replacing the words "org", "business" with "team". So I apologize for confusing the team to a for-profit org.

Could you still clarify the usual conditions under which people get a license for commercial usage and what the real reasons were to allow noncommercial but potentially unethical usages like pranking people who know the person who was a source of a voice, but disallow commercial but ethical usages, like talking machines? Does the team require commercial users to pay or barter something? I mean it is important to know if the team de-facto benefits from nonexistance of the docs on creating voices.

**Comment by abitrolly** [comment]
@KOLANICH if you're not interested in researching how to build voices yourself, go to https://github.com/RHVoice/RHVoice/discussions for your question about how unethical are people for not creating these voices for you. :D

**Comment by KOLANICH** [comment]
I don't say it is unethical not to create voices for me. It is completely ethical not to do the things that contradict to one's best interests. Such as working for free. Or such as helping competitors. 

I just want the team to explicitly declare its interests so we can better understand what we can and should expect from the team members and what we shouldn't.

**Comment by abitrolly** [comment]
@KOLANICH you yourself can't even explain why you can't just go and ask the offtopic question in the appropriate place, so I don't know why you would expect a consensus from several people on the matters they probably never thought about.

**Comment by KOLANICH** [comment]
>why you would expect a consensus from several people on the matters they probably never thought about.

I don't think so. The decision to license voices under any license (maybe except WTFPL) requires a consensus among all the holders. The decision to license under a non-free license ... requires a deeper consensus and maybe even a legally binding aggreement, because it is weird when the holder himself is not allowed to do something with something he has created.

**Comment by abitrolly** [comment]
@KOLANICH so why you can't just go and ask the offtopic question in the appropriate place?

**Comment by KOLANICH** [comment]
Because it is not an offtopic here, since it is related to this issue.

```dot
nv [label="non-free voices"];
fv [label="free voices"];
t [label="the Team"];
co [label="Free voices creators"];
d [label="public docs"];
s [label="know-how"];
cu [label="commercial users"];

t -> nv -> cu [color="green"];
co -> fv -> cu [color="green"];

nv -> fv [color="red"];
fv -> nf [color="red"];

cu -> t [color="gold"];  // What is this effect?
t -> cu [color="brown"]; // What is this effect?

cu -> co [color="cyan"]; 

d -> co -> fv [color="green"];

d -> t -> nv [color="green"];
s -> t -> nv [color="green"];
```

**Comment by abitrolly** [comment]
> Because it is not an offtopic here, since it is related to this issue.

Everything is related. Say in one sentence what is not offtopic, and why it is directly related to this issue, and I may believe you.

**Comment by zvezdochiot** [comment]
Moral dilemma: 
- [ ] allow to profit from the engine, thereby stimulating its development 
- [ ] insist on adherence to norms, thereby inhibiting development 

---

- [x] what is your choice?

**Comment by maniyax** [comment]
Hello, @KOLANICH!

1. We don't owe you anything. Please remember this.
2. Creating voice is very long and resource-intensive process. We didn't get any help from you or anyone else.
3. We (our team and speakers) are against uncontrolled commercial use of our products.
4. No one from RHVoice have rights to modify, use commercially, or make decisions on these voices, since you did not create them. This is entirely our prerogative.

I hope this question is settled?

**Comment by KOLANICH** [comment]
>We don't owe you anything. Please remember this.

I absolutely remember it. Even clarifying the situation in your message was an act of goodwill. Thank you for that. 

>Creating voice is very long and resource-intensive process. We didn't get any help from you or anyone else.

I know. I just wanna try to create a voice myself.

>We (our team and speakers) are against uncontrolled commercial use of our products.
>This is entirely our prerogative.

Thank you for explicitly clarifying the matter by writing this. That's what I expected to see.

So you try to achieve control, so availability of free (under free licenses) voices created by third parties will allow people to use RHVoice without your control, so it is not in the best interests of the team to provide any help to the ones who want to make free voices, so we should not expect members of the team to contribute to the docs on creating voices.

The knowledge of this fact increases probability of someone else creating the needed docs. Thank you for not using the [policy of deliberate ambiguity](https://en.wikipedia.org/wiki/Policy_of_deliberate_ambiguity).

**Comment by abitrolly** [comment]
@KOLANICH how do you earn money for living? I am sure if you can persuade your sponsors to support RHVoice case (and they'll be interested, because you've already spent so much time here that you could spend with them..) then maybe nobody will object to invest time to releasing a voice without commercial restrictions.

However, if you investigate the history of voice synthesizers and how companies set prices for blind users, you will probably understand how complicated economics around it are. Only true hackers can solve such problems.

**Comment by KOLANICH** [comment]
>However, if you investigate the history of voice synthesizers and how companies set prices for blind users, you will probably understand how complicated economics around it are.

I don't own any company that has set prices for blind users. I don't work for any company related to selling products tagetting blind users.

<details>
<summary>Offtop</summary>
And I am completely ignorant of the issues that blind users have. I cannot even imagine how one can walk outdoors while being blind and without help of other people or a dog or a special device (are there any?). How do they detect steps? How do they avoid cyclists (riding over sidewalks, and usually hasty Yandex Food couriers) and users of electric personal vehicles (even not being blind I have collided several times with electric scooter users)?

Зрячий незрячего не разумеет, в общем. Наоборот тоже верно.
</offtop>

>how do you earn money for living?

Work hard.

>I am sure if you can persuade your sponsors to support RHVoice case (and they'll be interested, because you've already spent so much time here that you could spend with them..)

It's not their business for what I spend my time. Neither yours. My employer currently is not interested in TTS software at all. It was I who is interested in TTS software because though I am not (by 2021-03-16) visually impaired, I know where I can apply it.

>then maybe nobody will object to invest time to releasing a voice without commercial restrictions.

I don't know enough about creating voices to even understand if I can do it and if I can how much time and energy it would take.

**Comment by joshknnd1982** [comment]
can I use windows10 to record sentences for a script for a new english voice, and then compile it into a sapi5 voice? Or is linux required at this time?

**Comment by KOLANICH** [comment]
I have briefly read the scripts within https://github.com/RHVoice/RHVoice/tree/master/src/scripts/general , nothing within them seems to be unusable on windows

**Comment by abitrolly** [comment]
Technically speaking, the PRAAT and Python are cross-platform, so it is a little effort to make them run on Windows. The problem is that we don't know which scripts to run to process recorded sentences.

So the answer is, yes, you can use any windows10 software to record sentences. They could be converted to any other format no problem, but we don't know which format is needed, and then how to compile them into voice.

**Comment by grzezlo** [comment]
I'd like to add Polish language to RHVoice, so obvious step is starting from adding new voice for language already existing.
After some precious hints from @beqabeqa473 and spending countless hours on experiments, I managed to compile needed tools under Linux, build a voice on original HTS Demo CMU arctic SLT, and nearly successfully generate voice for RHVoice.

I wrote a detailed tutorial documenting whole process.

This tutorial  is divided into 3 parts:

- Part 1. Installation of common tools.
- Part 2. Running original HTS demo CMU ARCTIC SLT, to be sure, that everything works as it should.
- Part 3. Generating the voice for RHVoice.

Part 1 and 2 are complete, but some important questions related to RHVoice, are still open.
I hope that software author or people experienced in voice creation help to answer these questions, so I will be able to complete the tutorial, and finally add it to the original repo.

At the moment text is in my fork at:
[Creating new voice](https://github.com/grzezlo/RHVoice/blob/gz-new-voice-tutorial/doc/en/Creating-new-voice.md).

Help and comments are welcome.

**Comment by abitrolly** [comment]
@grzezlo I need more time to read it fully, but I already can tell that this one is a very awesome article. Will be a very interesting reading. Thanks!

**Comment by KOLANICH** [comment]
@grzezlo, thank you for your lot of work!

@beqabeqa473, thank you too, I was positively surprised :+1: and I should admit that it turned out I was **not** completely right in my earlier opinion

>so it is not in the best interests of the team to provide any help to the ones who want to make free voices

**Comment by grzezlo** [comment]
Main problem solved. It was written, but ignored by me earlier.
hts22_bindir in training.cfg should contain path to binaries from HTS 2.2, not 2.3. HHed from this path is used onlyonce in voice-building-scripts, in the export-voice step.
After this change rhvoice-test worked, and after copying by hand rhvoice/data/myvoice/* to `nvda\addons\RHVoice-voice-English-Myvoice\data` I can hear "my voice" voice of the synthesizer.
Tomorrow I'll update tutorial, and as an interesting homework after this lesson, I'll try to generate voice from LJ speech dataset.

**Comment by grzezlo** [comment]
I just pushed     v0.2 of the tutorial:
    HTS2.2 and 2.3 parallell installation, proper configuration, and successful voice build.
    Some questions still open (see introduction).

[Read more...](https://github.com/grzezlo/RHVoice/blob/gz-new-voice-tutorial/doc/en/Creating-new-voice.md)

**Comment by julianfairfax** [comment]
> I'd like to add Polish language to RHVoice, so obvious step is starting from adding new voice for language already existing.
> After some precious hints from @beqabeqa473 and spending countless hours on experiments, I managed to compile needed tools under Linux, build a voice on original HTS Demo CMU arctic SLT, and nearly successfully generate voice for RHVoice.
> 
> I wrote a detailed tutorial documenting whole process.
> 
> This tutorial  is divided into 3 parts:
> 
> - Part 1. Installation of common tools.
> - Part 2. Running original HTS demo CMU ARCTIC SLT, to be sure, that everything works as it should.
> - Part 3. Generating the voice for RHVoice.
> 
> Part 1 and 2 are complete, but some important questions related to RHVoice, are still open.
> I hope that software author or people experienced in voice creation help to answer these questions, so I will be able to complete the tutorial, and finally add it to the original repo.
> 
> At the moment text is in my fork at:
> [Creating new voice](https://github.com/grzezlo/RHVoice/blob/gz-new-voice-tutorial/doc/en/Creating-new-voice.md).
> 
> Help and comments are welcome.
> 

Thank you for putting in the effort to make this! I may at some point attempt to follow it and add my own voice in English and in French (maybe also in German if the required training words aren't too complicated and if an accent is okay, as I'm still learning it).

Currently the selection of English voices leaves some stuff to be desired. I have settled on SLT for now, but it still sounds robotic and somewhat depressing. And the fact that a TTS engine has no French or German support makes it quite a bit less useful in Europe
